multisense_ros/src/multisense.cpp

```diff
@@ -0,0 +1,1651 @@
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Carnegie Robotics, LLC nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL CARNEGIE ROBOTICS, LLC BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+#include <arpa/inet.h>
+#include <chrono>
+#include <fstream>
+#include <sstream>
+
+#include <nlohmann/json.hpp>
+#include <opencv2/opencv.hpp>
+
+#include <Eigen/Geometry>
+#include <sensor_msgs/image_encodings.hpp>
+
+#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
+
+#include <multisense_ros/multisense.h>
+
+#include <MultiSense/MultiSenseSerialization.hh>
+#include <MultiSense/MultiSenseUtilities.hh>
+
+namespace lms = multisense;
+
+using namespace std::chrono_literals;
+
+namespace multisense_ros {
+
+namespace { // anonymous
+
+struct ParameterStep
+{
+    double offset = 0.0;
+    double delta = 0.0;
+};
+
+bool is_modern_camera(const lms::MultiSenseInfo::DeviceInfo::HardwareRevision &camera_type)
+{
+    return (camera_type == lms::MultiSenseInfo::DeviceInfo::HardwareRevision::S27 ||
+            camera_type == lms::MultiSenseInfo::DeviceInfo::HardwareRevision::S30 ||
+            camera_type == lms::MultiSenseInfo::DeviceInfo::HardwareRevision::KS21 ||
+            camera_type == lms::MultiSenseInfo::DeviceInfo::HardwareRevision::KS21_SILVER ||
+            camera_type == lms::MultiSenseInfo::DeviceInfo::HardwareRevision::KS21i);
+}
+
+std::chrono::microseconds get_closest_exposure(uint32_t exposure_us, uint32_t min, double step)
+{
+    return std::chrono::microseconds{static_cast<uint32_t>(std::floor((std::floor(static_cast<double>(exposure_us - min) / step) * step) + min))};
```

Comment:
unsure if step is guaranteed to be non zero somehow, but may want 0 check if returning 0 from the fn would be bad